src/c/coap-server.c

```diff
@@ -0,0 +1,465 @@
+  {
+    len = lib_addr->size = ainfo->ai_addrlen;
+    switch (ainfo->ai_family)
+    {
+    case AF_INET6:
+      memcpy (&lib_addr->addr.sin6, ainfo->ai_addr, lib_addr->size);
+      goto finish;
+    case AF_INET:
+      memcpy (&lib_addr->addr.sin, ainfo->ai_addr, lib_addr->size);
+      goto finish;
+    default:
+      ;
+    }
+  }
+
+ finish:
+  freeaddrinfo (res);
+  return len;
+}
+
+/* Caller must free returned iot_data_t */
+static iot_data_t*
+read_data_float64 (uint8_t *data, size_t len)
+{
+  /* data conversion requires a null terminated string */
+  uint8_t data_str[FLOAT64_STR_MAXLEN+1];
+  memset (data_str, 0, FLOAT64_STR_MAXLEN+1);
+  memcpy (data_str, data, len);
+
+  char *endptr;
+  errno = 0;
+  double dbl_val = strtod ((char *)data_str, &endptr);
+
+  if (errno || (*endptr != '\0'))
+  {
+    iot_log_info (sdk_ctx->lc, "invalid float64 of len %u", len);
+    return NULL;
+  }
+
+  return iot_data_alloc_f64 (dbl_val);
+}
+
+/* Caller must free returned iot_data_t */
+static iot_data_t*
+read_data_int32 (uint8_t *data, size_t len)
+{
+  /* data conversion requires a null terminated string */
+  uint8_t data_str[INT32_STR_MAXLEN+1];
+  memset (data_str, 0, INT32_STR_MAXLEN+1);
+  memcpy (data_str, data, len);
+
+  char *endptr;
+  errno = 0;
+  long int_val = strtol ((char *)data_str, &endptr, 10);
+
+  /* validate strtol conversion, and ensure within range */
+  if (errno || (*endptr != '\0') || (int_val < INT32_MIN) || (int_val > INT32_MAX))
+  {
+    iot_log_info (sdk_ctx->lc, "invalid int32 of len %u", len);
+    return NULL;
+  }
+
+  return iot_data_alloc_i32 ((int32_t) int_val);
+}
+
+/* Caller must free returned iot_data_t */
+static iot_data_t*
+read_data_string (uint8_t *data, size_t len)
+{
+  /* must copy request data to append null terminator */
+  char *str_data = malloc (len + 1);
+  memcpy (str_data, data, len);
+  str_data[len] = '\0';
+
+  iot_data_t *iot_data = iot_data_alloc_string(str_data, IOT_DATA_COPY);
+  free (str_data);
+
+  return iot_data;
+}
+
+/*
+ * Parse URI path, expect 3 segments: /a1r/{device-name}/{resource-name}
+ *
+ * @param[in] request For path to parse
+ * @param[out] device Found device
+ * @param[out] resource Found resource for device
+ * @return true if URI format OK, and device and resource found 
+ */
+static bool
+parse_path (coap_pdu_t *request, edgex_device **device_ptr, edgex_deviceresource **resource_ptr)
+{
+  coap_string_t *uri_path = coap_get_uri_path (request);
+  iot_log_debug (sdk_ctx->lc, "URI %s", uri_path->s);
+  char *path = (char *)uri_path->s;
+
+  edgex_device *device = NULL;
+  edgex_deviceprofile *profile = NULL;
+  edgex_deviceresource *resource = NULL;
+  
+  char *seg = strtok (path, "/");
+  bool res = false;
+  for (int i = 0; i < 3; i++)
+  {
+    if (!seg)
+    {
+      iot_log_info (sdk_ctx->lc, "missing URI segment %u", i);
+      break;
+    }
+
+    switch (i)
+    {
+    case 0:
+      if (strcmp (seg, RESOURCE_SEG1))
```

Comment:
The resource value is purposefully hardcoded here such that it always has to start with "a1r"? Assume a sensor device wants to send sensor data using coap to EdgeX. The sensor device uses /temperature as the resource name. In this case I cannot use device-coap service to receive data right? I should be able to configure the device-coap-c service with any resource name through the configuration file.  Please clarify.

Reply:
> The resource value is purposefully hardcoded here such that it always has to start with "a1r"?

Yes, purposefully hardcoded. The device-coap-c service is modeled after the [device-rest-go](https://github.com/edgexfoundry/device-rest-go) service, which hardcodes `/api/v1/resource` as the URL prefix. I used `/a1r` to be consistent with the often constrained MTU size of devices that run CoAP.

I think extending this implementation to customize the URL prefix would best be left to a follow-on PR. IMO it's outside the original scope of the service.